### ISSUE TITLE
Promote Instructor Center Card and List to egghead-ui

### ISSUE THREAD

**Issue by trevordmiller** [eggheadio/egghead-instructor-center#182]
The `List` component is currently implemented in the Instructor Center:
https://github.com/eggheadio/egghead-instructor-center/blob/master/src/components/List/index.js

All it does is accept a list of items, and renders them stacked vertically with A) a light border on the bottom of every item but the last and 2) equal padding around each item. It is often used inside of a `Card` component directly.

egghead-rails is now using `List`:
https://github.com/eggheadio/egghead-rails/pull/1211/files
so it could be promoted and consumed from egghead-ui.

## Examples from mocks

`List` was first designed in the Instructor Center mocks from @vojtaholik:

![image](https://cloud.githubusercontent.com/assets/5497885/24075477/f97ba06e-0be1-11e7-8118-ca42126b813c.png)

![image](https://cloud.githubusercontent.com/assets/5497885/24075479/01c21424-0be2-11e7-9cea-a34e2f71e72b.png)

![image](https://cloud.githubusercontent.com/assets/5497885/24075481/078d828a-0be2-11e7-93b1-ef5e09d41062.png)

![image](https://cloud.githubusercontent.com/assets/5497885/24075483/147c6966-0be2-11e7-921e-e2c3feb67206.png)

## Examples currently in Instructor Center

![image](https://cloud.githubusercontent.com/assets/5497885/24075495/4bbd8ac2-0be2-11e7-9abd-a059a7285490.png)

![image](https://cloud.githubusercontent.com/assets/5497885/24075498/55474056-0be2-11e7-9726-2fe970e991e4.png)

![image](https://cloud.githubusercontent.com/assets/5497885/24075503/5ebe93aa-0be2-11e7-8af2-2c93d3c485da.png)

**Comment by joelhooks** [comment]
Would like to see this type of story fleshed out more completely specified as a proposal for component API inclusion into the egghead-ui project. This will open an opportunity for comments and discussion versus a simple cut-n-paste of what has been implemented already. 

as an aside, I'd see the above level of detail on an issue as perfect for a "note" as a placeholder until time was available to specify the task more completely.

**Comment by trevordmiller** [comment]
@joelhooks Ok. I added this as an issue as it seemed straightforward and ready to be implemented to me. I've updated this issue's details ^ with more information and screenshots. I feel it is ready to go, but if you want to discuss more or make changes, let me know.

**Comment by joelhooks** [comment]
For inclusion in the egghead-ui project, will the list have any additional formatting options for things like color etc to make it more flexible?  

The name `List` seems generic, and could describe many things in our apps. This list has specific formatting, orientation, and other properties baked in to the implementations. Will we keep the generic name, or will the egghead-ui version be more specifically named to be in line with current implementations?

**Comment by trevordmiller** [comment]
I think it would be cool to use this generic simple list as long as we can for consistency and ease of use. I'm a big fan of components with minimal config required to ease the mental burden and ensure design consistency. But if we have other color/formatting needs it could be easily extended when that time comes. Unless there is a need for these things now, I don't see a reason to add extra functionality until it is needed. The reason why I named it a generic `List` is because it is being used for each "list" of things in the Instructor Center mocks; I'd assume that will continue with future mocks, but if there are going to be different types of lists, we could rename it to something more specific now or when the time comes. What do you think @vojtaholik? Does this list seem like a good candidate for the style guide now, or does it need to be more flexible now?

**Comment by trevordmiller** [comment]
Didn't mean to tap the close button

**Comment by joelhooks** [comment]
We've got another list that is currently being worked on as a generic component for egghead.io. It is being pulled from here: https://github.com/eggheadio/egghead-rails/blob/master/client/app/screens/Browse/screens/Tag/components/LessonsList/index.js#L95-L98

The design specification for this list component is here: https://egghead.invisionapp.com/share/A8AHXR0MN#/220017179_-__List_--_Component

Will its design require additional functionality that should be accounted for now?

**Comment by trevordmiller** [comment]
My suggestion would be to promote the `List` component as suggested in this PR in it's simple form (item padding and bottom border only). Then specifically in egghead-rails or any project, you can compose `List` with any children and extra behavior you want it to create other types of specific lists; for example, for the LessonsList use case mentioned, you could do:

```
import React from 'react'
import {map} from 'lodash'
import {List} from 'egghead-ui'
import LessonListItem from './components/LessonListItem'

const LessonList = ({lessons}) => (
  <List items={map(lessons, lesson => (
    <LessonListItem lesson={lesson} />
  ))} />
)

export default LessonList
```

Where `LessonListItem` looks like this (without its own padding):
![image](https://cloud.githubusercontent.com/assets/5497885/24076623/1240ab58-0bfb-11e7-9a59-f2614bab6b19.png)
And to get the hover enlargement effect, this component could have [a CSS `scale` transition](https://codepen.io/wifeo/pen/qzwkb).

Then using a `LessonList` in egghead-rails would look like this:
![image](https://cloud.githubusercontent.com/assets/5497885/24076628/38f18c36-0bfb-11e7-96d0-c35b5ceb5ca4.png)

This way, `List` is as minimal and single-responsibility as possible - all it does is add padding and a bottom border to each item in a list; the children are up to the consumer.

Does that make sense? I prefer simple components like this for a component library that can be composed together rather than giving them a bunch of formatting props and config.

**Comment by joelhooks** [comment]
That makes sense. We have these rounded corners on everything too, is that something the list takes care of? Internally the little "row cards" would style their own hover state, but I'm wondering more about the container they sit in, where it rounds the corners on the top and bottom item.

That visual pattern is all over the place, so I'm wondering who gets the job of controlling the corners?

**Comment by joelhooks** [comment]
![image](https://cloud.githubusercontent.com/assets/86834/24076705/c480217e-0bf4-11e7-9174-5e58b76b978d.png)

This is an example of the rounding, where we have the rows, but this List also has a "topper".

This might be some clever css class on list items too, I don't know, just wondering :)

**Comment by trevordmiller** [comment]
@joelhooks

In the instructor center, the `Card` component owns the border radius and drop shadow styling, list has none of that itself. So a `List` is often is put inside a `Card`.

So IMO this is how all these things should be composed together:

- [`Card`](https://github.com/eggheadio/egghead-instructor-center/blob/master/src/components/Card/index.js) is promoted to egghead-ui; it only concerns itself with the rounded corners and drop shadow)
- [`List`](https://github.com/eggheadio/egghead-instructor-center/blob/master/src/components/List/index.js) is promoted to egghead-ui ; it only concerns itself with borders and padding of items
- `LessonList` only exists in `egghead-rails` and composes together `Card` and `List` with egghead-rails specific sub-components and logic like this:

```
import React from 'react'
import {map} from 'lodash'
import {Card, List} from 'egghead-ui'
import PlayCourse from './components/PlayCourse'
import LessonListItem from './components/LessonListItem'

const LessonList = ({lessons, course}) => (
  <Card>
    {course
      ? <PlayCourse />
      : null
    }
    <List items={map(lessons, lesson => (
      <LessonListItem lesson={lesson} />
    ))} />
  </Card>
)

export default LessonList
```

![lesson-list-composition-example](https://cloud.githubusercontent.com/assets/5497885/24076924/da1f4772-0c02-11e7-8bd3-488cb3559000.png)

In summary: promote generic, single responsibility components in egghead-ui; then compose these things together with project specific logic and sub-components to create screens.

Here are a couple examples of this in the Instructor Center:
- https://github.com/eggheadio/egghead-instructor-center/blob/master/src/screens/Instructors/screens/Instructor/components/InstructorInfoCard/index.js
- https://github.com/eggheadio/egghead-instructor-center/blob/master/src/screens/Dashboard/components/HelpCard/index.js

Make sense?

**Comment by trevordmiller** [comment]
Closing, moving to new egghead-ui GitHub project: https://github.com/eggheadio/egghead-ui/issues/73. Will start using this to discuss in the "Icebox" column potential component/utility promotion as a team, then we can move to "Ready to implement" when ready.
